### PR TITLE
Introduce a `MessageGroup.condition`

### PR DESCRIPTION
**Pull request by artembilan**
* Add a `MessageGroup.condition` option
* Add a `MessageGroupStore.conditionSupplier` option
* Use it from the `SimpleMessageStore.addMessagesToGroup()` API
to populate `condition` (if any) into a `MessageGroup`
* Introduce a `GroupConditionProvider` contract to be implemented
on those `ReleaseStrategy` contracts which could be aware of group condition
* Populated a `GroupConditionProvider.getGroupConditionSupplier()`
into a `MessageGroupStore` from the `AbstractCorrelatingMessageHandler`
for end-user convenience
* Rework a `FileMarkerReleaseStrategy` to implement a `GroupConditionProvider`
to provide a function which produces a condition from `file_lineCount` header
of the `END` marker message
* Make the `FileMarkerReleaseStrategy` logic already based on the condition from a group
* Delegate `GroupConditionProvider` from the `FileAggregator`
* Add test for empty file aggregation

### CODE REVIEW

**Comment by artembilan** [comment]
Some PoC for `condition` in the `MessageGroup` and its usage from the `ReleaseStrategy`.

TODO: condition property implementation for every single `MessageGroupStore` impl...

**Comment by artembilan** [comment]
I think all the work is done with source code.
Only docs are left: I can add them to this PR or we can deffer them into the next milestone.
Or we even may not include this feature to the current milestone as well.

@garyrussell , let me know what works best for you to review this and have it merged one or another way.

**Comment by garyrussell** [comment]
I suggest at this late hour that we defer to the next milestone.

**Comment by artembilan** [comment]
> I suggest at this late hour that we defer to the next milestone.

Too late, I have just added docs, so it is ready for the final review 😉 

But yeah, we definitely can push it back if it doesn't work for you at the moment.

Thanks

**Comment by garyrussell** [comment]
It's too big to review; don't you want to get the builds under way ASAP?

File: spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java [hunk 1]
```diff
@@ -407,6 +407,11 @@ protected void onInit() {
 		 */
 		this.lockRegistrySet = true;
 		this.forceReleaseProcessor = createGroupTimeoutProcessor();
+
+		if (this.releaseStrategy instanceof GroupConditionProvider) {
+			this.messageStore.setConditionSupplier(
+					((GroupConditionProvider) this.releaseStrategy).getGroupConditionSupplier());
+		}
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java:414]: I don't think this will work - while we recommend a different store for each aggregator, we don't enforce it.

**Comment by artembilan** [comment]
Sure! No problem! Reviewing your PR for TCP/IP problem and that's it.

**Review by garyrussell — COMMENTED** [review]
Need an answer to my comment before reviewing further.

**Comment by artembilan** [comment]
Moved a `conditionSupplier` from the `MGS` to the `AbstractCorrelatingMessageHandler`, so we don't suffer on the `MGS` any more mutating other groups accidentally.

File: spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java [hunk 2]
```diff
@@ -294,6 +295,15 @@ public void removeMessageGroup(Object groupId) {
 		}
 	}
 
+	@Override
+	public void setGroupCondition(Object groupId, String condition) {
+		MessageGroupMetadata metadata = getGroupMetadata(groupId);
+		if (metadata != null) {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:301]: Perhaps we need an assertion here, to ensure `setGroupCondition` is only called after the current record has been added to the group (and metadata will be present).

File: spring-integration-core/src/main/resources/org/springframework/integration/config/spring-integration.xsd [hunk 3]
```diff
@@ -4071,6 +4071,19 @@
...
 				</xsd:attribute>
+				<xsd:attribute name="group-condition-supplier" type="xsd:string">
+					<xsd:annotation>
+						<xsd:appinfo>
+							<tool:annotation kind="ref">
+								<tool:expected-type type="java.util.function.BiFunction"/>
+							</tool:annotation>
+						</xsd:appinfo>
+						<xsd:documentation>
+							A reference to an 'java.util.function.BiFunction' bean that may provide a group
```
Comment (garyrussell) [inline spring-integration-core/src/main/resources/org/springframework/integration/config/spring-integration.xsd:4082]:
```suggestion
							A reference to a 'java.util.function.BiFunction' bean that may provide a group
```

File: spring-integration-jdbc/src/main/resources/org/springframework/integration/jdbc/schema-db2.sql [hunk 4]
```diff
@@ -18,7 +18,7 @@ CREATE TABLE INT_GROUP_TO_MESSAGE  (
 CREATE TABLE INT_MESSAGE_GROUP  (
 	GROUP_KEY CHAR(36) NOT NULL,
 	REGION VARCHAR(100) NOT NULL,
-	MARKED BIGINT,
+    CONDITION VARCHAR(255),
```
Comment (garyrussell) [inline spring-integration-jdbc/src/main/resources/org/springframework/integration/jdbc/schema-db2.sql:21]: We'll need a migration guide entry for the schema change.

File: spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/store/MessageDocument.java [hunk 5]
```diff
@@ -151,6 +154,15 @@ public Object getGroupId() {
...
 
+	@Nullable
+	public String getCondition() {
+		return this.condition;
+	}
+
+	public void setCondition(String condition) {
+		this.condition = condition;
+	}
+
```
Comment (garyrussell) [inline spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/store/MessageDocument.java:165]: Are there any backwards compatibility issues (existing documents)?

File: src/reference/asciidoc/message-store.adoc [hunk 6]
```diff
@@ -147,4 +147,21 @@ However starting with version 5.5, all the persistent `MessageGroupStore` implem
...
 A returned `Stream<Message<?>>` must be closed in the end of processing, e.g. via auto-close by the `try-with-resources`.
-Whenever a `PersistentMessageGroup` is used, its `streamMessages()` delegates to the `MessageGroupStore.streamMessagesForGroup()`.
\ No newline at end of file
+Whenever a `PersistentMessageGroup` is used, its `streamMessages()` delegates to the `MessageGroupStore.streamMessagesForGroup()`.
+
+[[message-group-condition]]
+==== Message Group Condition
+
+Starting with version 5.5, the `MessageGroup` abstraction provides a `condition` string option.
+The value of this option can be anything what could be parsed later on for any reason to make a decision for the group.
```
Comment (garyrussell) [inline src/reference/asciidoc/message-store.adoc:156]:
```suggestion
The value of this option can be anything that could be parsed later on for any reason to make a decision for the group.
```

File: src/reference/asciidoc/message-store.adoc [hunk 7]
```diff
@@ -147,4 +147,21 @@ However starting with version 5.5, all the persistent `MessageGroupStore` implem
...
+For example a `ReleaseStrategy` from a <<./aggregator.adoc#aggregator-api, correlation message handler>> may consult this property from the group instead of iterating all the messages in the group.
+The `MessageGroupStore` expose a `setGroupCondition(Object groupId, String condition)` API.
```
Comment (garyrussell) [inline src/reference/asciidoc/message-store.adoc:158]:
```suggestion
The `MessageGroupStore` exposes a `setGroupCondition(Object groupId, String condition)` API.
```

File: src/reference/asciidoc/message-store.adoc [hunk 8]
```diff
@@ -147,4 +147,21 @@ However starting with version 5.5, all the persistent `MessageGroupStore` implem
...
+For this purpose a `setGroupConditionSupplier(BiFunction<Message<?>, String, String>)` option has been added to the `AbstractCorrelatingMessageHandler`.
+This function is evaluated against each message to be added to the group and existing condition on the group.
```
Comment (garyrussell) [inline src/reference/asciidoc/message-store.adoc:160]:
```suggestion
This function is evaluated against each message after it has been added to the group as well as the existing condition of the group.
```

File: src/reference/asciidoc/message-store.adoc [hunk 9]
```diff
@@ -147,4 +147,21 @@ However starting with version 5.5, all the persistent `MessageGroupStore` implem
...
+The implementation may decide to return new value, existing or reset the target condition to `null`.
```
Comment (garyrussell) [inline src/reference/asciidoc/message-store.adoc:161]:
```suggestion
The implementation may decide to return a new value, the existing value, or reset the target condition to `null`.
```

File: src/reference/asciidoc/message-store.adoc [hunk 10]
```diff
@@ -147,4 +147,21 @@ However starting with version 5.5, all the persistent `MessageGroupStore` implem
...
+The value for a `condition` can be a JSON, SpEL expression, number or anything what can be serialized as a string and parsed afterwards.
+For example, the `FileMarkerReleaseStrategy` from the <<./file.adoc#file-aggregator, File Aggregator>> component, populates a condition into a group from the `FileHeaders.LINE_COUNT` header of the `FileSplitter.FileMarker.Mark.END` message and consults with it from its `canRelease()` comparing a group size with the value in this condition.
+This way it doesn't iterate all the messages in group to find a `FileSplitter.FileMarker.Mark.END` message with the `FileHeaders.LINE_COUNT` header.
```
Comment (garyrussell) [inline src/reference/asciidoc/message-store.adoc:164]: It also allows the end marker to arrive at the aggregator before all the other records; for example when processing a file in a multi-threaded environment.

File: src/reference/asciidoc/message-store.adoc [hunk 11]
```diff
@@ -147,4 +147,21 @@ However starting with version 5.5, all the persistent `MessageGroupStore` implem
...
+
+In addition, for target configuration convenience, a `GroupConditionProvider` contract has been introduced.
+The `AbstractCorrelatingMessageHandler` checks provided `ReleaseStrategy` for this contract implementation and extracts a `conditionSupplier` for further group condition evaluation logic.
```
Comment (garyrussell) [inline src/reference/asciidoc/message-store.adoc:167]:
```suggestion
In addition, for configuration convenience, a `GroupConditionProvider` contract has been introduced.
The `AbstractCorrelatingMessageHandler` checks if the provided `ReleaseStrategy` implements this interface and extracts a `conditionSupplier` for group condition evaluation logic.
```

File: src/reference/asciidoc/whats-new.adoc [hunk 12]
```diff
@@ -41,8 +41,12 @@ This is covered as a `ConsumerEndpointSpec.reactive()` option in Java DSL and as
 See <<./reactive-streams.adoc#reactive-streams,Reactive Streams Support>> for more information.
 
 The `groupTimeoutExpression` for a correlation message handler (an `Aggregator` and `Resequencer`) can now be evaluated to a `java.util.Date` for some fine-grained scheduling use-cases.
+Also  `BiFunction groupConditionSupplier` option is added to the `AbstractCorrelatingMessageHandler` to supply a `MessageGroup` condition against a message to be added to the group.
```
Comment (garyrussell) [inline src/reference/asciidoc/whats-new.adoc:44]: Also the

**Review by garyrussell — CHANGES_REQUESTED** [review]
Looks good.

File: spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java (hunk 2 above)
Comment (artembilan) [reply spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:301]: I don't think it is relevant. It is only the fact of an aggregator that we add a record then calculate a condition.
In other cases it might be set independently and unconditionally.

File: spring-integration-jdbc/src/main/resources/org/springframework/integration/jdbc/schema-db2.sql (hunk 4 above)
Comment (artembilan) [reply spring-integration-jdbc/src/main/resources/org/springframework/integration/jdbc/schema-db2.sql:21]: Sure! Will address it after merge

File: spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/store/MessageDocument.java (hunk 5 above)
Comment (artembilan) [reply spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/store/MessageDocument.java:165]: Existing documents comes without a `condition` field: nothing to read - nothing to map.
It is really a trick of all of these NO-SQL DBs: no schema - can change data structure any time!

File: spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java (hunk 2 above)
Comment (garyrussell) [reply spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:301]: ok

**Review by garyrussell — APPROVED** [review]

**Comment by artembilan** [comment]
See also a Migration Guide note: https://github.com/spring-projects/spring-integration/wiki/Spring-Integration-5.4-to-5.5-Migration-Guide#messagegroupconditon-option
